collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,162 @@
+                    processQueuedSamples()
+                }
+            })
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            if (retrySamplesSet.size > 0) {
+                val nextScheduledTime = retrySamplesSet.first().scheduledTime
+
+                if (retryToken != nextScheduledTime) {
+                    if (retryToken != null) {
+                        handler.removeCallbacks(processSampleRunnable, retryToken)
+                    }
+                    retryToken = nextScheduledTime
+                    val delay = maxOf(nextScheduledTime.time - Date().time, 0) // to prevent negative delay
+                    val t = handler.postAtTime(processSampleRunnable, retryToken, SystemClock.uptimeMillis() + delay)
+                    Log.d(TAG, t.toString())
+                }
+            }
+        } catch (e: Exception) {
+            Log.d(TAG, "processQueuedSamples exception ${e.message}")
+        }
+    }
+
+    private val processSampleRunnable = Runnable {
+        try {
+            val retrySample = getSample()
+            Log.d(TAG, "pop sample ${retrySample?.eventData?.sequenceNumber}")
+            if (retrySample != null) {
+                scheduleSample(retrySample)
+            }
+            retryToken = null
+            processQueuedSamples()
+        } catch (e: Exception) {
+            Log.d(TAG, "processSampleRunnable exception ${e.message}")
+        }
+    }
+
+    fun addSample(retrySample: RetrySample) {
+
+        try {
+            lock.lock()
+            retrySample.retry++
+            val backOffTime = minOf(2.toDouble().pow(retrySample.retry).toInt(), 64)
+            // more than 5min in queue
+            if (retrySample.totalTime + backOffTime < MAX_RETRY_TIME) {
+
+                retrySample.scheduledTime = Calendar.getInstance().run {
+                    add(Calendar.SECOND, backOffTime)
+                    time
+                }
+                retrySample.totalTime += backOffTime
+
+                if (retrySamplesSet.size > MAX_RETRY_SAMPLES) {
+                    val removeSample = retrySamplesSet.last()
+                    retrySamplesSet.remove(removeSample)
+                    Log.d(TAG, "removed retry sample ${removeSample?.eventData?.sequenceNumber} ")
+                }
+
+                Log.d(TAG, "add sample backOffTime=${retrySample.totalTime} scheduledTime=${retrySample.scheduledTime}")
+                retrySamplesSet.add(retrySample)
+
+                processQueuedSamples()
+            }
+        } catch (e: Exception) {
+            Log.d(TAG, "addSample exception ${e.message}")
+        } finally {
+            lock.unlock()
+        }
+    }
+
+    private fun getSample(): RetrySample? {
+        try {
+            lock.lock()
+            val retrySample = retrySamplesSet.firstOrNull { it.scheduledTime.before(Date()) || it.scheduledTime == Date() }
```

Comment:
```suggestion
            val retrySample = retrySamplesSet.firstOrNull { it.scheduledTime.compareTo(Date()) <= 0 }
```
or `>=`, not sure now, should also do the trick and is easier to read.